src/util.js

```diff
@@ -107,3 +107,80 @@ export function format_memory_and_limit(usage, limit) {
         return _("");
     }
 }
+
+export function updateContainers() {
+    let newContainers = {};
+    let newContainersStats = {};
+    let newContainersMeta = [];
+    let inspectRet = [];
+    let containerStatsRet = [];
+    let countRun;
+    return new Promise((resolve, reject) => {
+        varlinkCall(PODMAN, "io.podman.ListContainers")
+                .then(reply => {
+                    newContainersMeta = reply.containers;
+                    if (!newContainersMeta) {
+                        console.log(newContainersMeta);
+                        resolve({newContainers: newContainers, newContainersStats: newContainersStats});
+                    }
+                    inspectRet = newContainersMeta.map((container) => {
+                        return new Promise((resolve, reject) => {
+                            varlinkCall(PODMAN, "io.podman.InspectContainer", {name: container.id})
+                                    .then(reply => {
+                                        resolve(JSON.parse(reply.container));
+                                    })
+                                    .catch(ex => reject(new Error("Failed to do InspectContainer call:", ex, JSON.stringify(ex))));
+                        });
+                    });
+                })
+                .then(function() {
```

Comment:
I find it hard to mentally follow what's going on in this "ladder" of multiple chained `.then()`s. The block above  does not by itself return a promise, so I believe this isn't really a chain that waits on subsequent promises, but all of these `.then()` blocks run directly after one another.

I think it would make things much easier to understand if all of that code runs in a single `.then()` handler of the `ListContainers` call. (For the record, you are allowed to use empty lines to set apart different steps in the code, to make it easier to read ☺)

Also, style nitpick: Please use `function` or `=>` consistently. I suggest to use `() =>` here; in general, arrow functions are what you want in such promise handlers, as it has a much more intuitive semantics of `this`.

Reply:
I was wondering if using only one .then(), the functions inside then() will execute not in sequence and app.jsx can not get the complete data, it is possible?
```js
new Promise(() => {
    ListContainers()
        .then() {
            InspectContainer();
            getStats();
            resolve()
        }
})
```

Reply:
It's actually the other way around. In every imperative programming language, statements in a function are always done in sequence - JavaScript is no different.

With multiple `.then()`s, there is the possibility of chaining them or running them in parallel, that's where things get complicated (see https://javascript.info/promise-chaining). But  we don't need that here.